lang/c/src/schema.c

```diff
@@ -556,6 +560,77 @@ avro_schema_record_field_append(const avro_schema_t record_schema,
 	return 0;
 }
 
+
+static int
+avro_schema_record_field_append_with_default_json(const avro_schema_t record_schema,
+                                                  const char *field_name,
+                                                  const avro_schema_t field_schema,
+                                                  const json_t *field_default)
+{
+	check_param(EINVAL, is_avro_schema(record_schema), "record schema");
+	check_param(EINVAL, is_avro_record(record_schema), "record schema");
+	check_param(EINVAL, field_name, "field name");
+	check_param(EINVAL, is_avro_schema(field_schema), "field schema");
+        check_param(EINVAL, field_default, "field_default");
+
+	if (!is_avro_id(field_name)) {
+		avro_set_error("Invalid Avro identifier");
+		return EINVAL;
+	}
+
+	if (record_schema == field_schema) {
+		avro_set_error("Cannot create a circular schema");
+		return EINVAL;
+	}
+
+	struct avro_record_schema_t *record = avro_schema_to_record(record_schema);
+	struct avro_record_field_t *new_field = (struct avro_record_field_t *) avro_new(struct avro_record_field_t);
+	if (!new_field) {
+		avro_set_error("Cannot allocate new record field");
+		return ENOMEM;
+	}
+	new_field->index = record->fields->num_entries;
+	new_field->name = avro_strdup(field_name);
+	new_field->type = avro_schema_incref(field_schema);
+        new_field->default_value = json_deep_copy((json_t *) field_default);
+
+        /*
+         * TODO: Do we need to validate the default value types here?  If I have
+         * a json_t object, should I use `json_typeof` or the `json_is_` functions
+         * in Jannson to verify the json_type and field_schema are the same? How
+         * far down do we go for complex types?
```

Comment:
Yes we need to validate that the JSON is a valid instance of the field schema.  An internal helper method (or a handful of them) will probably make this easier.